c_src/sasl_auth.c

```diff
@@ -129,13 +128,15 @@ static int load(ErlNifEnv* env, void** UNUSED(priv), ERL_NIF_TERM UNUSED(info))
     ATOM_UNKNOWN = enif_make_atom(env, "unknown");
     ATOM_NOT_CONTROLLING_PROCESS = enif_make_atom(env, "not_controlling_process");
 
-    sasl_resource = init_resource_type(env);
-    int result;
     sasl_set_mutex((sasl_mutex_alloc_t*)&sasl_mutex_alloc, (sasl_mutex_lock_t*)&sasl_mutex_lock,
         (sasl_mutex_unlock_t*)&sasl_mutex_unlock, (sasl_mutex_free_t*)&sasl_mutex_free);
 
-    result = sasl_client_init(NULL);
-    return !sasl_resource && !(result == SASL_OK);
+    sasl_client_connection_nif_resource_type = init_resource_type(env, "sasl_auth_cli_state");
+    int cli_result = sasl_client_init(NULL);
+    sasl_server_connection_nif_resource_type = init_resource_type(env, "sasl_auth_srv_state");
+    int srv_result = sasl_server_init(NULL, "sasl_auth");
+    return !sasl_client_connection_nif_resource_type && !(cli_result == SASL_OK)
+        && !sasl_server_connection_nif_resource_type && !(srv_result == SASL_OK);
```

Comment:
when server state failed to init, it should release client resource before return?

Reply:
there was no enif_release_resource for the resource type though
maybe both should be released on unload ?

Reply:
```
static ErlNifResourceType* init_resource_type(ErlNifEnv* env, const char* name)
{
    return enif_open_resource_type(env, NULL, name, (ErlNifResourceDtor*)destroy_resource,
        ERL_NIF_RT_CREATE | ERL_NIF_RT_TAKEOVER, NULL);
}
```
Already passed the `destroy_resource` callback(destructor function) in the enif_open_resource_type. don't need to destroy_resource on unload.

> typedef void ErlNifResourceDtor(ErlNifEnv* caller_env, void* obj);
>
>The function prototype of a resource destructor function.
>The obj argument is a pointer to the resource. The only allowed use for the resource in the destructor is to access its user data one final time. The destructor is guaranteed to be the last callback before the resource is deallocated.